Standalone Compiler Updates

alxhub opened the pull request:
This PR bundles together a number of commits that round out the implementation of standalone components/directives/pipes in the AOT compiler and linker.

alxhub committed:
refactor(core): change component emit to 'dependencies'

Previously, the compiler would represent template dependencies of a
component in its component definition through separate fields (`directives`,
`pipes`).

This commit refactors the compiler/runtime interface to use a single field
(`dependencies`). The runtime component definition object still has separate
`directiveDefs` and `pipeDefs`, which are calculated from the `dependencies`
when the definition is evaluated.

This change is also reflected in partially compiled declarations. To ensure
compatibility with partially compiled code already on NPM, the linker
will still honor the old form of declaration (with separate fields).

alxhub committed:
refactor(compiler-cli): unify tracked template scope dependencies

Previously, the compiler tracked directives and pipes in template scopes
separately. This commit refactors the scope system to unify them into a
single data structure, disambiguated by a `kind` field.

alxhub committed:
fix(compiler-cli): fix issue with incremental tracking of APIs for pipes

`PipeSymbol` contains logic to detect changes in the public API surface of
pipes, which includes the pipe name. However, the pipe handler inadvertently
uses the pipe class name instead of the actual pipe name to initialize the
`PipeSymbol`, which breaks incremental compilation when pipe names change.

There is a test which attempts to verify that this logic is working, but the
test actually passes for a different reason. The test swaps the names of 2
pipes that are both used in a component, and asserts that the component is
re-emitted, theoretically because the public APIs of the pipes is changed.
However, the emit order of the references to the pipes depends on the order
in which they match in the template, which changes when the names are
swapped. This ordering dependency is picked up by the semantic dependency
tracking system, and is what actually causes the component to be re-emitted
and therefore the pipe test to pass in spite of the bug with name tracking.

This commit fixes the `PipeSymbol` initialization to use the correct pipe
name. The test is still flawed in that it's sensitive to the ordering of
pipe emits, but this ordering is due to change soon as a result of the
standalone components work, so this issue will be resolved in a future
commit.

alxhub committed:
refactor(compiler-cli): emit NgModule dependencies of standalone components

This commit expands on the unified dependency tracking in the previous
commit and adds tracking of NgModule dependencies. These are not used for
standard components, but are emitted for standalone components to allow the
runtime to roll up providers from those NgModules into standalone injectors.

alxhub committed:
refactor(compiler-cli): emit StandaloneFeature for standalone components

This commit adds an emit for standalone components of the
`StandaloneFeature`, which will support creation of standalone injectors and
any other mechanisms necessary for standalone component functionality at
runtime.

Using a feature allows for standalone functionality to be tree-shaken in
applications that aren't using them.

alxhub committed:
refactor(compiler-cli): move standalone scope computation out of handler

Standalone component scopes were first implemented in the
`ComponentDecoratorHandler` itself, due to an assumption that "standalone"
allowed for a localized analysis of the component's dependencies. However,
this is not strictly true. Other compiler machinery also needs to understand
component scopes, including standalone component scopes. A good example is
the template type-checking engine, which uses a `ComponentScopeReader` to
build full metadata objects (that is, metadata that considers the entire
inheritance chain) for type-checking purposes. Therefore, the
`ComponentScopeReader` should be able to give the scope for a standalone
component.

To achieve this, a new `StandaloneComponentScopeReader` is implemented, and
the return type of `ComponentScopeReader.getScopeForComponent` is expanded
to express standalone scopes. This cleanly integrates the "standalone"
concept into the existing machinery.

alxhub committed:
test(compiler-cli): tests for standalone components/directives/pipes

This commit bundles tests for standalone components that are possible after
previous implementation commits. Most new tests are compliance tests, but
a test is also included to validate that the template type-checking system
can work with standalone components as well.

alxhub committed:
refactor(compiler-cli): emit isStandalone flag in partial declarations

This commit propagates the `isStandalone` flag for a component, directive,
or pipe during partial compilation of a standalone declaration. This flag
allows the linker to properly process a standalone declaration that it
encounters.

pkozlowski-opensource commented on packages/compiler/src/render3/partial/directive.ts:
```diff
@@ -51,6 +51,10 @@ export function createDirectiveDefinitionMap(meta: R3DirectiveMetadata):
   // e.g. `type: MyDirective`
   definitionMap.set('type', meta.internalType);
 
+  if (meta.isStandalone) {
```
So we keep having this discussion about the naming of this field (`isStandalone` vs. `standalone`, similar to the existing boolean flag to identify pure pipes).

WDYT about settling on the `standalone ` name?

alxhub replied:
The compiler internally uses `isStandalone`, as part of a larger convention (we also have `isComponent`, `isPoisoned`, etc).

pkozlowski-opensource commented on packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:
```diff
@@ -550,14 +561,14 @@ export class ComponentDecoratorHandler implements
 
+    // const scope = scopeTemplate(
+    //     this.scopeReader, this.dtsScopeReader, this.scopeRegistry, this.metaReader, node,
+    //     analysis, this.usePoisonedData);
```
nit: remove the commented out code if no longer needed?

atscott commented on packages/compiler-cli/src/ngtsc/scope/src/standalone.ts:
```diff
@@ -0,0 +1,113 @@
+      private dtsModuleReader: DtsModuleScopeResolver) {}
+
+  getScopeForComponent(clazz: ClassDeclaration): StandaloneScope|null {
+    if (!this.cache.has(clazz)) {
```
```suggestion
    if (this.cache.has(clazz)) {
      return this.cache.get(clazz)!;
    }

alxhub replied:
This is the typical pattern for memoized computations in the compiler. It's basically:
```
if (!cache.has(key)) {
  const result = computeThing();
  cache.set(key, result);
}
return cache.get(key)!;
```

This structure helps avoid a common pitfall of returning the result _without_ putting it in the cache, since there's one clear return point and it always reads from the canonical cached location.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -198,7 +204,12 @@ export interface R3DeclareComponentMetadata extends R3DeclareDirectiveMetadata {
+    R3DeclareDirectiveDependencyMetadata|R3DeclarePipeDependencyMetadata;
+
+export interface R3DeclareDirectiveDependencyMetadata {
+  kind: 'directive'|'component';
```
FYI, we use symbolic references to enum types in other places, e.g. the target type of a factory function:

```
target: i0.ɵɵFactoryTarget.Component
```

The current approach with string literals does avoid the need for a stable symbol so is perhaps more desirable, but I'm mentioning it here if we want to unify these for consistency.

alxhub replied:
Good to know - I think I'll keep the string constants for now as it's more reliable.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -149,15 +149,21 @@ export interface R3DeclareComponentMetadata extends R3DeclareDirectiveMetadata {
+   * List of dependencies which matched in the template, including sufficient
+   * metadata for each
+   */
+  dependencies?: R3DeclareTemplateDependencyMetadata[];
```
We should arguably bump the `minVersion` of components because of this change (in a followup commit/PR is fine).

alxhub replied:
👍  agreed - and for directives/pipes due to the need to compile the `standalone` flag.

JoostK commented on packages/compiler/src/render3/view/api.ts:
```diff
@@ -254,15 +245,35 @@ export interface R3ComponentMetadata extends R3DirectiveMetadata {
+
   /**
-   * The type of the directive as an expression.
+   * The type of the dependency as an expression;
```
The `;` seems strange here

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:
```diff
@@ -602,91 +602,113 @@ export class ComponentDecoratorHandler implements
       const cyclesFromPipes = new Map<UsedPipe, Cycle>();
-      for (const usedPipe of usedPipes) {
-        const cycle = this._checkForCyclicImport(usedPipe.importedFile, usedPipe.type, context);
+      for (const usedDep of declarations) {
```
In the future, I believe we should be able to skip the cyclic dep check for standalone components as any reference we may need to generate an import from is already part of the import graph from the standalone component.

alxhub replied:
Agreed - I need to think through the logic fully but I believe you're right and we can skip it.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -249,6 +249,12 @@ export interface R3DeclarePipeDependencyMetadata {
 }
 
+export interface R3DeclareNgModuleDependencyMetadata {
+  kind: 'ngmodule';
```
My preference would be `ngModule` here.

JoostK commented on packages/compiler/src/render3/partial/component.ts:
```diff
@@ -173,6 +173,12 @@ function compileUsedDependenciesMetadata(meta: R3ComponentMetadata<R3TemplateDep
+        const ngModuleMeta = new DefinitionMap<R3DeclareNgModuleDependencyMetadata>();
+        ngModuleMeta.set('kind', o.literal('ngmodule'));
+        ngModuleMeta.set('type', wrapType(decl.type));
+        // TODO: implement this.
```
What's the followup action here?

alxhub replied:
To remove the TODO, of course!

JoostK commented on packages/compiler-cli/src/ngtsc/scope/src/typecheck.ts:
```diff
@@ -85,11 +85,16 @@ export class TypeCheckScopeRegistry {
-    if (this.scopeCache.has(scope.ngModule)) {
-      return this.scopeCache.get(scope.ngModule)!;
+    const cacheKey = scope.kind === ComponentScopeKind.NgModule ? scope.ngModule : scope.component;
+    const dependencies = scope.kind === ComponentScopeKind.NgModule ?
+        scope.compilation.dependencies :
+        scope.dependencies;
```
I'd move this below checking the cache to make it more apparent when this is actually needed.

JoostK commented on packages/compiler-cli/src/ngtsc/scope/src/standalone.ts:
```diff
@@ -0,0 +1,113 @@
+        return null;
+      }
+
+      // A standalone component always has itself in scope, so add `clazzMeta` during
+      // initialization.
```
Noting this here; it's a bit unclear what the status of the JIT compilation is at this point, but we should ensure that we account for this situation in JIT as well.

JoostK reviewed (approved):
LGTM overall. We should not forget about the `.d.ts` emit side of standalone components.

AndrewKushnir commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -149,15 +149,21 @@ export interface R3DeclareComponentMetadata extends R3DeclareDirectiveMetadata {
+  directives?: R3DeclareDirectiveDependencyMetadata[];
+
+  /**
+   * List of dependencies which matched in the template, including sufficient
+   * metadata for each
```
nit: looks like an incomplete sentence.

AndrewKushnir commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -226,6 +237,24 @@ export interface R3DeclareUsedDirectiveMetadata {
+export interface R3DeclareNgModuleDependencyMetadata {
+  kind: 'ngmodule';
+
+  type: o.Expression|(() => o.Expression);
```
nit: may be copy the field description from the `R3DeclarePipeDependencyMetadata` above for completeness?

AndrewKushnir commented on packages/compiler/src/render3/partial/component.ts:
```diff
@@ -156,46 +150,35 @@ function computeEndLocation(file: ParseSourceFile, contents: string): ParseLocat
+        const ngModuleMeta = new DefinitionMap<R3DeclareNgModuleDependencyMetadata>();
+        ngModuleMeta.set('kind', o.literal('ngmodule'));
+        ngModuleMeta.set('type', wrapType(decl.type));
+        // TODO: implement this.
```
Leftover comment or more updates are needed?

AndrewKushnir reviewed (approved):
Awesome work @alxhub 👍 

Just left a few minor comments, none of them are blocking and up to you to address or ignore (I'm ok either way :)).

AndrewKushnir reviewed (approved):
Reviewed-for: size-tracking

pkozlowski-opensource reviewed (approved):
Going over this PR again - it LGTM. While my understanding of the compiler is limited I also don't see any "red flags" here. 

Again, the breakdown into small, focused commits is great and helps a lot! Thnx @alxhub !

alxhub committed:
refactor(compiler-cli): propagate standalone flag in .d.ts metadata

This commit adds a type field to .d.ts metadata for directives, components,
and pipes which carries a boolean literal indicating whether the given type
is standalone or not. For backwards compatibility, this flag defaults to
`false`.

Tests are added to validate that standalone types coming from .d.ts files
can be correctly imported into new standalone components.

alxhub commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -407,37 +409,66 @@ function convertOpaqueValuesToExpressions(obj: {[key: string]: OpaqueValue}):
+
+function convertDirectiveDeclarationToMetadata(
+    declaration: R3DeclareDirectiveDependencyFacade,
+    isComponent: true|null = null): R3DirectiveDependencyMetadata {
```
This is actually pretty specifically designed - passing `false` for an `isComponent` parameter is a clear statement that "no, this is not a component." I would expect `isComponent = false` to override any other sources of component-ness.

Using `null` communicates "there's no information about whether this is a component", so the caller can either say "yes it's a component" or "idk if it's a component, you figure it out".

AndrewKushnir replied:
Sounds good. Can we copy'n'paste your comment into the code? :) Not a blocker though, feel free to ignore.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:
```diff
@@ -584,14 +589,20 @@ export class ComponentDecoratorHandler implements
-        }
+      const pipes = new Map<string, PipeMeta>();
+
+      let {dependencies, isPoisoned} =
+          scope.kind === ComponentScopeKind.NgModule ? scope.compilation : scope;
+      if (scope.kind === ComponentScopeKind.NgModule) {
+        dependencies = scope.compilation.dependencies;
       }
```
You're absolutely right.

atscott reviewed (approved):
reviewed-for: size-tracking

alxhub committed:
test: set skipLibCheck on ngcc integration test

The ngcc integration test is in an awkward state: it's attempting to test
that the current ngcc can process @angular/core at v12. We need to make a
forwards-incompatible change to the typings of @angular/core, which means
that the compiled typings from the current ngcc won't be compatible with
core as of v12.

To get around this and allow the integration test to have some value, we're
disabling library checking for the time being.

alxhub commented:
Caretaker: this PR doesn't actually affect the public API, it changes the shape of Angular metadata on compiled classes which is reflected in the .d.ts. It doesn't actually need public-api review.

dylhunn commented:
This PR needs a rebase on `master`

alxhub commented:
This PR was merged into the repository by commit b8d3389173f49e7c1fb1c799654c7f854a967007.
